src/objects.c

```diff
@@ -51,24 +53,118 @@ P11PROV_OBJ *p11prov_obj_new(P11PROV_CTX *ctx, CK_SLOT_ID slotid,
     obj->class = class;
+    obj->cached = CK_INVALID_HANDLE;
 
     obj->refcnt = 1;
 
     return obj;
 }
 
-P11PROV_OBJ *p11prov_obj_ref(P11PROV_OBJ *obj)
+static void destroy_key_cache(P11PROV_OBJ *obj, P11PROV_SESSION *session)
+
+{
+    P11PROV_SESSION *_session = NULL;
+    CK_SESSION_HANDLE sess;
+    CK_RV ret;
+
+    if (obj->cached == CK_INVALID_HANDLE) {
+        return;
+    }
+
+    if (session) {
+        sess = p11prov_session_handle(session);
+    } else {
+        ret = p11prov_take_login_session(obj->ctx, obj->slotid, &_session);
+        if (ret != CKR_OK) {
+            P11PROV_debug("Failed to get login session. Error %lx", ret);
+            return;
+        }
+        sess = p11prov_session_handle(_session);
+    }
+
+    ret = p11prov_DestroyObject(obj->ctx, sess, obj->cached);
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to destroy cached key. Error %lx", ret);
+    }
+    obj->cached = CK_INVALID_HANDLE;
+
+    if (_session) {
+        p11prov_return_session(_session);
+    }
+}
+
+static void cache_key(P11PROV_OBJ *obj)
+{
+    P11PROV_SESSION *session = NULL;
+    CK_BBOOL val_false = CK_FALSE;
+    CK_ATTRIBUTE template[] = { { CKA_TOKEN, &val_false, sizeof(val_false) } };
+    CK_SESSION_HANDLE sess;
+    CK_RV ret;
+
+    /* We cache only keys on the token */
+    if ((obj->class != CKO_PRIVATE_KEY && obj->class != CKO_PUBLIC_KEY)
+        || obj->cka_token != CK_TRUE) {
+        return;
+    }
+
+    ret = p11prov_take_login_session(obj->ctx, obj->slotid, &session);
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to get login session. Error %lx", ret);
+        return;
+    }
+
+    /* If already cached, release and re-cache */
+    destroy_key_cache(obj, session);
+
+    sess = p11prov_session_handle(session);
+    ret = p11prov_CopyObject(obj->ctx, sess, obj->handle, template, 1,
+                             &obj->cached);
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to cache key. Error %lx", ret);
+    }
+
+    P11PROV_debug("Key %lu:%lu cached as %lu:%lu", obj->slotid, obj->handle,
+                  session, obj->cached);
+
+    p11prov_return_session(session);
+    return;
+}
+
+P11PROV_OBJ *p11prov_obj_ref_no_cache(P11PROV_OBJ *obj)
 {
+    P11PROV_debug("Ref Object: %p (handle:%lu)\n", obj,
+                  obj ? obj->handle : CK_INVALID_HANDLE);
+
     if (obj && __atomic_fetch_add(&obj->refcnt, 1, __ATOMIC_SEQ_CST) > 0) {
         return obj;
     }
 
     return NULL;
 }
 
+P11PROV_OBJ *p11prov_obj_ref(P11PROV_OBJ *obj)
+{
+    obj = p11prov_obj_ref_no_cache(obj);
+    if (!obj) {
+        return NULL;
+    }
+
+    /* When referenced it means we are likely going to try to use the key in
+     * some operation, let's try to cache it in the tokens volatile memory for
+     * those tokens that support the operation. This will result in much faster
+     * key operations with some tokens as the keys are unencrypted in volatile
+     * memory */
```

Comment:
I would say this is very NSS-specific. See also the previous comment. I am not against adding this, but I would probably not enable doing this by default on all tokens and gate this behind some configuration option (or pkcs11 module detection/quirk)

Reply:
Will softhsm or pcscd suffer from doing this? (Performance, other?)
If not and there is no harm I would leave it on. Unclear how I would configure this, and explicit option in openssl.cnf ?

Reply:
not sure about softhsm, but this will end up in OpenSC here as unimplemented so some check to avoid calling this over and over again (or having quirk for opensc) would be needed anyway:
https://github.com/OpenSC/OpenSC/blob/70771735ae10180bb039043b9a1b00b66bf00fc1/src/pkcs11/pkcs11-object.c#L166-L173

Reply:
Alright, I will add a tunable to skip caching entirely if we get that

Reply:
Added quirk